# Practice 3

### Martyna Baran, Zuzanna Jarlaczńska

### Exercise 1

In [1]:
import pandas as pd
import numpy as np

In [2]:
Zoo_filename = "zoo.data"
columns = ['animal_name', 'hair','feathers', 'eggs', 'milk', 'airborne', 'aquatic', 'predator', 'toothed', 'backbone', 'breathes', 'venomous', 'fins', 'legs', 'tail', 'domestic', 'catsize', 'type']
zooData = pd.read_csv(Zoo_filename, names=columns)
zooData

animal_name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

Firstly I deleted the animal_name column as i believe it is not important, because we care about the group type of the aminals.

In [3]:
zooData = zooData.drop(['animal_name'], axis=1)

#### a) using Holdout 25%

In [4]:
from sklearn.model_selection import train_test_split
import random
from sklearn.dummy import DummyClassifier

X_zoo = zooData[zooData.columns[:-1]]
Y_zoo = zooData['type']

seed = random.randint(0, 100)
X_train, X_test, Y_train, Y_test = train_test_split(X_zoo, Y_zoo,test_size=0.25, random_state=seed,stratify=Y_zoo)

dummy_clf = DummyClassifier(strategy='most_frequent')
dummy_clf.fit(X_zoo, Y_zoo)
ypred = dummy_clf.predict(X_test)



In [5]:
from sklearn import metrics

print(metrics.classification_report(Y_test,ypred, zero_division=0))
print("Kappa metric: ", metrics.cohen_kappa_score(Y_test, ypred))

metrics.confusion_matrix(Y_test, ypred)

              precision    recall  f1-score   support

           1       0.42      1.00      0.59        11
           2       0.00      0.00      0.00         5
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         3
           5       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         2
           7       0.00      0.00      0.00         3

    accuracy                           0.42        26
   macro avg       0.06      0.14      0.08        26
weighted avg       0.18      0.42      0.25        26

Kappa metric:  0.0


array([[11,  0,  0,  0,  0,  0,  0],
       [ 5,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0],
       [ 1,  0,  0,  0,  0,  0,  0],
       [ 2,  0,  0,  0,  0,  0,  0],
       [ 3,  0,  0,  0,  0,  0,  0]], dtype=int64)

In [6]:
class_counts = zooData['type'].value_counts()
class_ratios = class_counts / len(zooData)
print("Class distribution in the Zoo Dataset")
print(class_ratios)

Class distribution in the Zoo Dataset
1    0.405941
2    0.198020
4    0.128713
7    0.099010
6    0.079208
3    0.049505
5    0.039604
Name: type, dtype: float64


In [7]:
class_counts = Y_test.value_counts()
class_ratios = class_counts / len(Y_test)
print("Class distribution in test set")
print(class_ratios)

Class distribution in test set
1    0.423077
2    0.192308
7    0.115385
4    0.115385
6    0.076923
3    0.038462
5    0.038462
Name: type, dtype: float64


In [8]:
class_counts = Y_train.value_counts()
class_ratios = class_counts / len(Y_train)
print("Class distribution thete train set")
print(class_ratios)

Class distribution thete train set
1    0.400000
2    0.200000
4    0.133333
7    0.093333
6    0.080000
3    0.053333
5    0.040000
Name: type, dtype: float64


The ZeroR is the procedure for classification algorithms whose output is simply the most frequently occurring classification in a set of data. As we can see in our data the most frequent class is class 1. Using Hold-out method we divide our dataset into train and test parts. What is important, we need to make sure that the distributions of the classes elements is the same (or very close) in train and test sets - and this is the responsibility of the stratify parameter of dividing procedure. Then our accuracy is very similar to the frequency of this class (as we made sure the ratios are similar in test and train sets). In our example class 1 gave accuracy of 42% as in is the number of its samples in the test set. Important thing is that we set the zero_division attribute to be equal 0 and because of that we have 0's in precision column where the n/0 division occured(I wanted to notice that because it will be the same in every exercise).  

##### b) repeted holdout

In [9]:
accuracies =[]

for i in range(30):
    seed = i*10
    X_train, X_test, y_train, y_test = train_test_split(X_zoo, Y_zoo, test_size=0.25, random_state=seed)
    dummy_clf.fit(X_train, y_train)
    y_pred = dummy_clf.predict(X_test)
    
    report = metrics.classification_report(y_test, y_pred, zero_division=0)
    print(f"Iteration {i+1}, seed {seed}:\n{report}")
    print(metrics.confusion_matrix(y_test, y_pred))
    
    acc= metrics.accuracy_score(y_test, y_pred)
    accuracies.append(acc)


accur = sum(accuracies)/30

print(f"Average accuracy is: {accur} and the standard deviation is {np.std(accuracies)}")



Iteration 1, seed 0:
              precision    recall  f1-score   support

           1       0.38      1.00      0.56        10
           2       0.00      0.00      0.00         5
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         5
           5       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         2
           7       0.00      0.00      0.00         2

    accuracy                           0.38        26
   macro avg       0.05      0.14      0.08        26
weighted avg       0.15      0.38      0.21        26

[[10  0  0  0  0  0  0]
 [ 5  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  0]
 [ 5  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  0]
 [ 2  0  0  0  0  0  0]
 [ 2  0  0  0  0  0  0]]
Iteration 2, seed 10:
              precision    recall  f1-score   support

           1       0.35      1.00      0.51         9
           2       0.00      0.00      0.00         4
           3       0.00     

Iteration 14, seed 130:
              precision    recall  f1-score   support

           1       0.35      1.00      0.51         9
           2       0.00      0.00      0.00         7
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         2
           5       0.00      0.00      0.00         2
           6       0.00      0.00      0.00         1
           7       0.00      0.00      0.00         4

    accuracy                           0.35        26
   macro avg       0.05      0.14      0.07        26
weighted avg       0.12      0.35      0.18        26

[[9 0 0 0 0 0 0]
 [7 0 0 0 0 0 0]
 [1 0 0 0 0 0 0]
 [2 0 0 0 0 0 0]
 [2 0 0 0 0 0 0]
 [1 0 0 0 0 0 0]
 [4 0 0 0 0 0 0]]
Iteration 15, seed 140:
              precision    recall  f1-score   support

           1       0.46      1.00      0.63        12
           2       0.00      0.00      0.00         2
           3       0.00      0.00      0.00         1
           4      

[[12  0  0  0  0  0]
 [ 4  0  0  0  0  0]
 [ 1  0  0  0  0  0]
 [ 4  0  0  0  0  0]
 [ 4  0  0  0  0  0]
 [ 1  0  0  0  0  0]]
Iteration 26, seed 250:
              precision    recall  f1-score   support

           1       0.38      1.00      0.56        10
           2       0.00      0.00      0.00         8
           3       0.00      0.00      0.00         2
           4       0.00      0.00      0.00         2
           6       0.00      0.00      0.00         1
           7       0.00      0.00      0.00         3

    accuracy                           0.38        26
   macro avg       0.06      0.17      0.09        26
weighted avg       0.15      0.38      0.21        26

[[10  0  0  0  0  0]
 [ 8  0  0  0  0  0]
 [ 2  0  0  0  0  0]
 [ 2  0  0  0  0  0]
 [ 1  0  0  0  0  0]
 [ 3  0  0  0  0  0]]
Iteration 27, seed 260:
              precision    recall  f1-score   support

           1       0.46      1.00      0.63        12
           2       0.00      0.00      0.00   

In [10]:
accuracies = []

for i in range(30):
    seed = random.randint(0,300)
    X_train, X_test, y_train, y_test = train_test_split(X_zoo, Y_zoo, test_size=0.25)
    dummy_clf.fit(X_train, y_train)
    y_pred = dummy_clf.predict(X_test)
    
    report = metrics.classification_report(y_test, y_pred, zero_division=0)
    print(f"Iteration {i+1}, seed {seed}:\n{report}")
    print(metrics.confusion_matrix(y_test, y_pred))
    acc= metrics.accuracy_score(y_test, y_pred)
    accuracies.append(acc)
    
    
accur = sum(accuracies)/30

print(f"Average accuracy is: {accur} and the standard deviation is {np.std(accuracies)}")

Iteration 1, seed 122:
              precision    recall  f1-score   support

           1       0.46      1.00      0.63        12
           2       0.00      0.00      0.00         5
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         5
           6       0.00      0.00      0.00         2
           7       0.00      0.00      0.00         1

    accuracy                           0.46        26
   macro avg       0.08      0.17      0.11        26
weighted avg       0.21      0.46      0.29        26

[[12  0  0  0  0  0]
 [ 5  0  0  0  0  0]
 [ 1  0  0  0  0  0]
 [ 5  0  0  0  0  0]
 [ 2  0  0  0  0  0]
 [ 1  0  0  0  0  0]]
Iteration 2, seed 24:
              precision    recall  f1-score   support

           1       0.54      1.00      0.70        14
           2       0.00      0.00      0.00         6
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         3
           7  

Iteration 17, seed 191:
              precision    recall  f1-score   support

           1       0.35      1.00      0.51         9
           2       0.00      0.00      0.00         5
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         4
           5       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         4
           7       0.00      0.00      0.00         2

    accuracy                           0.35        26
   macro avg       0.05      0.14      0.07        26
weighted avg       0.12      0.35      0.18        26

[[9 0 0 0 0 0 0]
 [5 0 0 0 0 0 0]
 [1 0 0 0 0 0 0]
 [4 0 0 0 0 0 0]
 [1 0 0 0 0 0 0]
 [4 0 0 0 0 0 0]
 [2 0 0 0 0 0 0]]
Iteration 18, seed 68:
              precision    recall  f1-score   support

           1       0.31      1.00      0.47         8
           2       0.00      0.00      0.00         6
           3       0.00      0.00      0.00         4
           4       

Iteration 30, seed 13:
              precision    recall  f1-score   support

           1       0.42      1.00      0.59        11
           2       0.00      0.00      0.00         8
           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         2
           5       0.00      0.00      0.00         1
           6       0.00      0.00      0.00         2
           7       0.00      0.00      0.00         1

    accuracy                           0.42        26
   macro avg       0.06      0.14      0.08        26
weighted avg       0.18      0.42      0.25        26

[[11  0  0  0  0  0  0]
 [ 8  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  0]
 [ 2  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  0]
 [ 2  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  0]]
Average accuracy is: 0.41538461538461546 and the standard deviation is 0.06691276077281995


The characteristic feature of Repeted Holdout method is the fact that we don't have ensurance that every class representant will appear in test set. Because of that we can obtained (sometimes very) different results as despite having in test set more samples from for example class 2 we still classify everything as class 1. 


In every iteration and every split of dataset we still got the division in which we classify everything to the most frequent class which is the first one. We have performed this algorithm twice: 
- with the seed value eqal number of iteation multiplied by 10 
- with random number from range (0,300) (which is almost the same as multiplying by 10).
We did it mostly because of curiosity of how the random_seed parameter influence the results.

In the first, the accuracy is in range 0.19 with seed 50 to 0.5 with seed 40 (so one after another). In both cases we have all the classes in test set. Pretty high results we also get with seeds 260 and 270. The precision always equals the accuracy - it's obvious having known that precision is the fraction of properly classified instances among all classified (our case properly classified is class 1) and the accuracy is the fraction of the class 1 instances in test set. 17th times we had all the classes in the test set.
In the second we got higher scores, because the best one was the 0.62 and the lowest 0.23. But with higher accuracy we didn't use all the classes (class 3 was skiped). The highest score with all the classes included is 0.46. All the classes were chosen 15th times.


Macro average is calculated as sum of scores of all the class divided by numer of class present in test set. 

Weigtened average is calculated as multipication of fracion of given class times the value of score for this class.

As we can see most of the times macro average has lower values because it treats every value equally and the Precision, Recall and F1-score are always zeros for class different than 1. In fact, weighted average also does not have very great results. 


Summing up, we calculated the average accuracy of the sets and despite higher results in the secon algorithm we get almost the same average. In both cases we got low standard deviation which is good, because it means that 

Repeted holdout method gave us higher possible result than the hold-out method. We can say that it might be a sightly better method as we can obtain results with different seeds. On the other hand the average accuracy was worse. 

So as it was the ZeroR algorithm we can say that every other algorithm should give us better results than 42-50% accuracy.

### Exercise 2

In [11]:
climate_filename = "pop_failures.dat"
cols = ['Study','Run','vconst_corr','vconst_2','vconst_3','vconst_4','vconst_5','vconst_7','ah_corr','ah_bolus','slm_corr','efficiency_factor','tidal_mix_max','vertical_decay_scale','convect_corr','bckgrnd_vdc1','bckgrnd_vdc_ban','bckgrnd_vdc_eq','bckgrnd_vdc_psim','Prandtl','outcome']
climateData = pd.read_csv(climate_filename, names=cols, sep='\s+')
climateData

Study  Run           vconst_corr             vconst_2  \
0    Study  Run           vconst_corr             vconst_2   
1        1    1    0.8590362064372231   0.9278245357602525   
2        1    2    0.6060410252371287   0.4577283627625244   
3        1    3    0.9975997770308621   0.3732384871005909   
4        1    4    0.7834078593063168   0.1040553139561477   
..     ...  ...                   ...                  ...   
536      3  176    0.6571360217355605   0.4893749949690472   
537      3  177    0.9158937899007772   0.8427200389133456   
538      3  178    0.4785995552678489   0.9411848060916075   
539      3  179  0.007792555536919585   0.7792874706236439   
540      3  180    0.6080745654943813  0.03155552789361941   

               vconst_3             vconst_4             vconst_5  \
0              vconst_3             vconst_4             vconst_5   
1    0.2528656223672442    0.298838311199668   0.1705212998954166   
2    0.3594484227670667   0.3069573767716065   0.8433307671777179   
3    0.5173993555298593   0.5049925456690189   0.6189033359884182   
4    0.1975326951075759   0.4218371592474998   0.7420556675966105   
..                  ...                  ...                  ...   
536  0.1337129038823251   0.4119495612103493  0.08777968800988876   
537  0.5189467905347961  0.09062169845371197   0.3369811584021793   
538  0.7692448065732607   0.9507760047175301   0.1894064358228611   
539  0.8674678949755616   0.7048200657474809   0.9832821685988974   
540  0.5982638455694541   0.7947711921749741   0.1456804811527642   

               vconst_7               ah_corr            ah_bolus  ...  \
0              vconst_7               ah_corr            ah_bolus  ...   
1     0.735936040913738    0.4283254278810799  0.5679469417703027  ...   
2    0.9348506611513181    0.4445724883608314  0.8280149253395697  ...   
3    0.6055708232180527    0.7462253300408419  0.1959282919790389  ...   
4    0.4908278824153564  0.005525436956021521  0.3921232672136588  ...   
..                  ...                   ...                 ...  ...   
536  0.3562886222770127    0.4802043909090571  0.0296780063990607  ...   
537   0.893576032403184    0.9787026870361945  0.6748679740315614  ...   
538  0.1127434391204992    0.7456450003031124  0.5270963716915705  ...   
539  0.4203031653719437    0.7106123455515546  0.1747464587976639  ...   
540  0.3781825262419362    0.4619484540825296  0.4252905171687922  ...   

      efficiency_factor       tidal_mix_max  vertical_decay_scale  \
0     efficiency_factor       tidal_mix_max  vertical_decay_scale   
1    0.2456748547110086  0.1042258652974851    0.8690907029810155   
2    0.6168699015410514  0.9757855814040846    0.9143436669954098   
3    0.6793550279741694  0.8034130759821791    0.6439951609962413   
4    0.4714626999390829  0.5978789030832963    0.7616587520758105   
..                  ...                 ...                   ...   
536  0.2805462690367778  0.3841168819254058    0.8859482824737724   
537    0.79810830750632  0.3535461527644657   0.04479587182868272   
538  0.1931025246093567  0.8295629554790341     0.101506036223792   
539  0.7611336392919638  0.4367138491141507    0.6901315704453737   
540  0.4809384926015304  0.3078155124391843    0.2316379503838511   

           convect_corr        bckgrnd_vdc1     bckgrnd_vdc_ban  \
0          convect_corr        bckgrnd_vdc1     bckgrnd_vdc_ban   
1    0.9975184956319734  0.4486200774528293  0.3075217871189428   
2    0.8452471419276359  0.8641518680676301  0.3467126889826937   
3    0.7184411332114703   0.924775074027841  0.3153714063448004   
4    0.3627505607476147  0.9128190944424003  0.9779711752522013   
..                  ...                 ...                 ...   
536  0.7684815695121264  0.4594791249928272  0.3344821448761246   
537  0.9908996256417595  0.3470274911055135  0.5124990522382885   
538  0.5488780750505006  0.3819661899030001  0.1988109947156368   
539  0.8251331410254352  0.9816564693245002

In [12]:
climateData= climateData.drop(0)

In [13]:
class_counts = climateData['outcome'].value_counts()
class_ratios = class_counts / len(climateData)
print(class_ratios)

1    0.914815
0    0.085185
Name: outcome, dtype: float64


In [16]:
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import KFold, cross_val_score, LeaveOneOut
from sklearn.neighbors import KNeighborsClassifier
import random

X = climateData[climateData.columns[2:-1]]
Y = climateData['outcome']
k=10
seed = random.randint(0,10)
minmaxSc = MinMaxScaler(feature_range=(0, 1))
minmaxSc.fit(X)
X_train_scaled = minmaxSc.transform(X)

print('K-fold cross-validation strategy')
for i in [2,6,10]:
    
    kfold = KFold(n_splits=k, random_state=seed, shuffle=True)
    model = KNeighborsClassifier(n_neighbors=i)
    res = cross_val_score(model, X_train_scaled, Y, cv=kfold, scoring = 'accuracy')
    print(f'Accuracy of the model with splits = {k}, random seed = {seed} and number of neighbors {i} is:\n {res}')
    print(f'Average accuracy is: {sum(res)/k}')
    
print('Leave One Out cross-validation strategy')    
for i in [2,6,10]:
    loocv = LeaveOneOut()
    model = KNeighborsClassifier(n_neighbors=i)
    res = cross_val_score(model, X_train_scaled, Y, cv=loocv)
    res = [str(int(x)) for x in res]
    print(f'Accuracy of the model with number of neighbours {i} is: {metrics.accuracy_score(Y, res)}')
    

K-fold cross-validation strategy
Accuracy of the model with splits = 10, randm seed = 10 and number of neighbours 2 is:
 [0.92592593 0.85185185 0.77777778 0.81481481 0.7962963  0.90740741
 0.90740741 0.92592593 0.85185185 0.83333333]
Average accuracy is: 0.8592592592592592
Accuracy of the model with splits = 10, randm seed = 10 and number of neighbours 6 is:
 [0.94444444 0.96296296 0.92592593 0.87037037 0.87037037 0.96296296
 0.94444444 0.94444444 0.88888889 0.94444444]
Average accuracy is: 0.9259259259259259
Accuracy of the model with splits = 10, randm seed = 10 and number of neighbours 10 is:
 [0.94444444 0.94444444 0.92592593 0.87037037 0.85185185 0.96296296
 0.96296296 0.94444444 0.88888889 0.92592593]
Average accuracy is: 0.9222222222222222
Leave One Out cross-validation strategy
Accuracy of the model with number of neighbours 2 is: 0.8703703703703703
Accuracy of the model with number of neighbours 6 is: 0.9666666666666667
Accuracy of the model with number of neighbours 10 is: 0.

Here we had very unbiased set: aproximately 91% of patterns were from class 1 and only 9% of class 2. As a result every model with accuracy below 91% is not good one. 

In the K-fold cross-validation strategy we got one with average accuracy only 85%, and the other two got around 92%. Having in mind the distribution of classes in out dataset those are rather poor results - it can mean that we simply classified all the patterns as class 1. 

On the other hand in LeaveOneOut we also get one result below 91%, but the other two got way better results! Especialy the third one that has 98% of accuracy - it is not perfect but in my opinion pretty high result. It was able to classify some of the class 2 patterns even if there were very small amount of them.

So in that particular case with my value of seed, k_splits and neighbors the Leave One Out method obtained better results and can be consider as a better choice. 

(PS. I believe that changing teh random seed could improve the performance of the K-fold method, but then we will lose the randomness of choice)

In [23]:
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import KFold, cross_val_score, LeaveOneOut
from sklearn.neighbors import KNeighborsClassifier
import random

X = climateData[climateData.columns[2:-1]]
Y = climateData['outcome']
k=5
#seed = random.randint(0,10)
seed = 10
minmaxSc = MinMaxScaler(feature_range=(0, 1))
minmaxSc.fit(X)
X_train_scaled = minmaxSc.transform(X)

print('K-fold cross-validation strategy')
for i in [2,6,10]:
    
    kfold = KFold(n_splits=k, random_state=seed, shuffle=True)
    model = KNeighborsClassifier(n_neighbors=i)
    res = cross_val_score(model, X_train_scaled, Y, cv=kfold, scoring = 'accuracy')
    print(f'Accuracy of the model with splits = {k}, random seed = {seed} and number of neighbors {i} is:\n {res}')
    print(f'Average accuracy is: {sum(res)/k}')
    
print('Leave One Out cross-validation strategy')    
for i in [2,6,10]:
    loocv = LeaveOneOut()
    model = KNeighborsClassifier(n_neighbors=i)
    res = cross_val_score(model, X_train_scaled, Y, cv=loocv)
    res = [str(int(x)) for x in res]
    print(f'Accuracy of the model with number of neighbours {i} is: {metrics.accuracy_score(Y, res)}')
    

K-fold cross-validation strategy
Accuracy of the model with splits = 5, random seed = 10 and number of neighbors 2 is:
 [0.90740741 0.78703704 0.83333333 0.91666667 0.85185185]
Average accuracy is: 0.8592592592592594
Accuracy of the model with splits = 5, random seed = 10 and number of neighbors 6 is:
 [0.9537037  0.89814815 0.90740741 0.94444444 0.90740741]
Average accuracy is: 0.9222222222222223
Accuracy of the model with splits = 5, random seed = 10 and number of neighbors 10 is:
 [0.94444444 0.89814815 0.88888889 0.9537037  0.90740741]
Average accuracy is: 0.9185185185185185
Leave One Out cross-validation strategy
Accuracy of the model with number of neighbours 2 is: 0.8703703703703703
Accuracy of the model with number of neighbours 6 is: 0.9666666666666667
Accuracy of the model with number of neighbours 10 is: 0.9814814814814815


Here I also wanted to perform this algorithm with k=5 number of splits and with the same seed to see how each parameter influences the results. As we can see they are very similar as before with k=10. We conclude that in this case this parameter do not have a lot of power, but maybe changing both number of splits and seed we would be able to perform better.

### Exercise 3

In [18]:
trainData = [
    [4.6, 3.2, 1.4, 1],
    [5.3, 3.7, 1.5, 3],
    [5.7, 4.4, 1.5, 1],
    [5, 3.5, 1.6, 2],
    [5.5, 2.5,4,1],
    [5.7, 3, 4.2, 2],
    [5.7, 2.8, 4.1, 2],
    [5.8, 2.7, 5.1, 1],
    [6.3, 2.5, 5, 2],
    [5.9, 3, 5.1, 3]
]
columns =['x1', 'x2', 'x3', 'class']
rows = ['1', '2', '3', '4', '5', '6', '7', '8','9', '10']
train = pd.DataFrame(trainData, columns=columns, index=rows)

X_train = train[train.columns[:-1]]
Y_train = train['class']

In [19]:
testData = [
    [5, 3.5, 1.7, 1],
    [4.3, 2.8, 1.5, 1],
    [2.7, 4.5, 1.2, 3],
    [5, 4.2, 1.3, 3],
    [6.3, 2.5, 4.1, 1],
    [5.2, 3, 4.5, 2],
    [4.5, 3, 4.2, 2],
    [5.9, 2.9, 5.2, 2],
    [5, 2.4, 5.1, 1],
    [4.5, 3.2, 5, 2]
]
    
test = pd.DataFrame(testData, columns=columns, index=rows)
X_test = test[test.columns[:-1]]
Y_test = test['class']

In [20]:
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

minmaxSc = MinMaxScaler(feature_range=(0, 1))
minmaxSc.fit(X_train)
X_train_scaled = minmaxSc.transform(X_train)
X_test_scaled = minmaxSc.transform(X_test)

In [21]:
for i in [2,5]:
    knn = KNeighborsClassifier(n_neighbors = i, metric='euclidean')
    knn.fit(X_train_scaled, Y_train)
    ypred = knn.predict(X_test_scaled)
    print(metrics.classification_report(Y_test, ypred, zero_division=0))
    print("Kappa value: ", metrics.cohen_kappa_score(Y_test, ypred))
    print(metrics.confusion_matrix(Y_test, ypred))

              precision    recall  f1-score   support

           1       0.33      0.50      0.40         4
           2       0.25      0.25      0.25         4
           3       0.00      0.00      0.00         2

    accuracy                           0.30        10
   macro avg       0.19      0.25      0.22        10
weighted avg       0.23      0.30      0.26        10

Kappa value:  -0.16666666666666674
[[2 2 0]
 [3 1 0]
 [1 1 0]]
              precision    recall  f1-score   support

           1       0.43      0.75      0.55         4
           2       0.67      0.50      0.57         4
           3       0.00      0.00      0.00         2

    accuracy                           0.50        10
   macro avg       0.37      0.42      0.37        10
weighted avg       0.44      0.50      0.45        10

Kappa value:  0.16666666666666652
[[3 1 0]
 [2 2 0]
 [2 0 0]]


From the table we can see that in both cases we didn't manage to predict the member of the third class. We had only two samples of third class so it was to little for our algorithm. Also, in both cases algorithms often precidted samples to be from class 1 than 2, despite having the same number of samples from those classes. The kappa value is very dissapointing - it means that our model performed very badly, as it compares obtained accuracy with expected accuracy which can be obtained by random classifier. But in the end, having in mind that is was very easy classification with only 10 samples I think that the accuracy of 50% in the second iteration with 5 neighbors is pretty good.